### ISSUE TITLE
AsyncDaemon does not restart after transient database problem

### ISSUE THREAD

**Issue by marcinbudny** [JasperFx/marten#1035]
**Marten version:** 2.8.0

**Problem**
After transient database connection problem, the daemon does not restart the projections. In fact, it seems to be stuck on stopping the projection.  [Repro can be found here.](https://gist.github.com/marcinbudny/59b8a6c60fbb4892e9590e5813d565c7)
When I restart my postgres database running in docker, the last thing that gets logged from daemon logger is:
```
Npgsql.NpgsqlException (0x80004005): Exception while reading from stream ---> System.IO.EndOfStreamException: A
ttempted to read past the end of the stream.
   at Npgsql.ReadBuffer.Ensure(Int32 count, Boolean async, Boolean dontBreakOnTimeouts)
   at Npgsql.ReadBuffer.Ensure(Int32 count, Boolean async, Boolean dontBreakOnTimeouts)
   at Npgsql.NpgsqlConnector.DoReadMessage(Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readin
gNotifications, Boolean isPrependedMessage)
   at Npgsql.NpgsqlConnector.ReadPrependedMessages(Boolean async)
   at Npgsql.NpgsqlConnector.ReadMessage(Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingN
otifications)
   at Npgsql.NpgsqlConnector.ReadExpecting[T](Boolean async)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async)
   at Npgsql.NpgsqlDataReader.<NextResultAsync>b__31_0()
   at Npgsql.NpgsqlCommand.Execute(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken
)
   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior, Boolean async, CancellationToken cance
llationToken)
   at Npgsql.NpgsqlCommand.<>c__DisplayClass90_0.<<ExecuteDbDataReaderAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Marten.Events.Projections.Async.Fetcher.buildEventPage(Int64 from, NpgsqlCommand cmd)
   at Marten.Events.Projections.Async.Fetcher.fetchNextPage(Int64 lastEncountered)
   at Marten.Events.Projections.Async.Fetcher.<>c__DisplayClass26_0.<<FetchNextPage>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Marten.Events.Projections.Async.ErrorHandling.DaemonErrorHandler.TryAction(Func`1 action, IMonitoredActiv
ity activity, Int32 attempts)
Stopping projection MartenDeamonProblem.CounterValue
```
Notice "stopping" but no "stopped". The daemon does not resume projections after database is up again.

**Expected behaviour**
The daemon should resume projections when database is reachable again.

**Possible cause of the problem**
I tried to track the problem and it seems to originate in the "pause" action in exception handling.

```csharp
private async Task pause(IMonitoredActivity activity)
{
	try
	{
		await activity.Stop().ConfigureAwait(false);
	}
	catch (Exception e)
	{
		_logger.Error(e);
	}

	await Task.Delay(_handling.Cooldown).ConfigureAwait(false);

	try
	{
		await activity.Start().ConfigureAwait(false);
	}
	catch (Exception e)
	{
		_logger.Error(e);
	}
}
```
The pause action waits for the `ProjectionTrack` to stop, which in turn waits for the `Fetcher` to stop. But `Fetcher` stop action is this:
```csharp
public Task Stop()
{
	if (State != FetcherState.Active)
	{
		return Task.CompletedTask;
	}

	State = FetcherState.Waiting;

	return _fetchingTask;
}
```
The fetching task will never complete, because it is this task that initiated the pause in the first place. So basically this task awaits itself to complete but at the same time prevents itself from completing.

If the pause code was changed to this:
```csharp
private Task pause(IMonitoredActivity activity)
{
	try
	{
		activity.Stop().ConfigureAwait(false);
	}
	catch (Exception e)
	{
		_logger.Error(e);
	}

	Task.Run(async () =>
	{
		await Task.Delay(_handling.Cooldown);
		try
		{
			await activity.Start().ConfigureAwait(false);
		}
		catch (Exception e)
		{
			_logger.Error(e);
		}
	});
	
	return Task.CompletedTask;
	
}
```
Then the daemon behaves as expected. But this is quite ugly solution. Any idea how to solve it in a better way?

**Comment by jeremydmiller** [comment]
I don't think the `Fetcher.Stop()` is a necessarily a problem unless it's a threading issue that deadlocks somewhere. Your code just kicks it into a different thread, right?

I don't have a better idea and there's not points for style here, so I'm going to use your fix in 2.9 and see how this goes for now

**Comment by marcinbudny** [comment]
Hi Jeremy,

The deadlock was exactly the problem I observed, with the task waiting for itself to complete but at the same time preventing itself from completing. 
The sample solution I posted is not very well tested. However the basic path worked: when postgres was down, the deamon was in a stop-pause-start loop and resumed when postgres was back online. Probably behavior in edge cases needs to be considered.
